core/optaplanner-constraint-streams-bavet/src/main/java/org/optaplanner/constraint/streams/bavet/bi/BavetFilterBiConstraintStream.java

```diff
@@ -68,39 +65,7 @@ public ConstraintStream getTupleSource() {
     @Override
     public <Score_ extends Score<Score_>> void buildNode(NodeBuildHelper<Score_> buildHelper) {
         buildHelper.<BiTuple<A, B>> putInsertUpdateRetract(this, childStreamList,
-                insert -> new ConditionalBiInserter<>(predicate, insert),
-                (update, retract) -> new ConditionalBiUpdater<>(predicate, update, retract));
-    }
-
-    private static final class ConditionalBiInserter<A, B> extends AbstractInserter<BiTuple<A, B>> {
-        private final BiPredicate<A, B> predicate;
-
-        public ConditionalBiInserter(BiPredicate<A, B> predicate, Consumer<BiTuple<A, B>> insert) {
-            super(insert);
-            this.predicate = predicate;
-        }
-
-        @Override
-        protected boolean test(BiTuple<A, B> tuple) {
-            return predicate.test(tuple.factA, tuple.factB);
-        }
-
-    }
-
-    private static final class ConditionalBiUpdater<A, B> extends AbstractUpdater<BiTuple<A, B>> {
-        private final BiPredicate<A, B> predicate;
-
-        public ConditionalBiUpdater(BiPredicate<A, B> predicate, Consumer<BiTuple<A, B>> update,
-                Consumer<BiTuple<A, B>> retract) {
-            super(update, retract);
-            this.predicate = predicate;
-        }
-
-        @Override
-        protected boolean test(BiTuple<A, B> tuple) {
-            return predicate.test(tuple.factA, tuple.factB);
-        }
-
+                tupleLifecycle -> new ConditionalBiTupleLifecycle<>(predicate, tupleLifecycle));
```

Comment:
No extra indirection here. ConditionalInsert + ConditionUpdate is replaced by ConditionalTupleLifecycle.

If anything, this will be better, because both insert and update now share the same `test(...)` method, helping JIT.